streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -528,7 +521,8 @@ private void maybeScheduleCheckpoint() {
 
     private void writeCheckpointIfNeed() {
         if (commitNeeded) {
-            throw new IllegalStateException("A checkpoint should only be written if no commit is needed.");
+            throw new IllegalStateException("A checkpoint should only be written if the previous commit has completed"
+                                                + " and there is no new commit needed.");
```

Comment:
`and there is no new commit needed` -> this seem to be miss leading because the `commitNeeded` flag is not really a guard for this case. -- Also, `if the previous commit has complete` is something we don't really know here.

Reply:
> and there is no new commit needed -> this seem to be miss leading because the commitNeeded flag is not really a guard for this case.

Isn't that exactly what the `commitNeeded` flag is?

That said, looking at this again, I agree this new phrasing doesn't really make sense. But the original comment took me a while to understand also (shouldn't we _only_ write a checkpoint if there's a commit needed?) IIUC the point is that we should not write a checkpoint before doing the actual commit, ie there should not be pending uncommitted data when we write the checkpoint.

Reply:
I'll revert it to the original and add a error log message: does
```log.error("Tried to write a checkpoint with pending uncommitted data, should complete the commit first.");```
make sense to you?